python/lsst/db/mysqlUtils.py

```diff
@@ -0,0 +1,387 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <http://www.lsstcorp.org/LegalNotices/>.
+
+"""
+This module contains utilities / helpers that perform common tasks related to
+MySQL. This includes:
+ * reading mysql credential files
+ * managing database and tables (create, drop, check if exists, listing, checking
+   if a table is a view)
+ * checking if user exists
+ * loading SQL scripts
+
+@author  Jacek Becla, SLAC
+"""
+
+
+# standard library
+import ConfigParser
+import logging as log
+import os.path
+import subprocess
+import sys
+
+# third party
+import sqlalchemy
+from sqlalchemy.exc import DBAPIError
+
+# local
+from lsst.db.db import Db, DbException
+
+
+# MySQL errors that we are catching
+# Names and numbers from include/mysql/mysql.h
+class MySqlErr:
+    ER_DB_CREATE_EXISTS = 1007
+    ER_DB_DROP_EXISTS = 1008
+    ER_NO_DB_ERROR = 1046
+    ER_BAD_DB_ERROR = 1049
+    ER_TABLE_EXISTS_ERROR = 1050
+    ER_BAD_TABLE_ERROR = 1051
+
+# Map of MySQLdb driver connect() keywords to mysql executable option names.
+# Note: 'read_default_group' is not supported, since it can cause
+# the MySQLdb driver and the mysql executable to connect differently.
+connectArgToOptionMap = {
+    'host':              'host',
+    'user':              'user',
+    'passwd':            'password',
+    'db':                'database',
+    'port':              'port',
+    'unix_socket':       'socket',
+    'connect_timeout':   'connect_timeout',
+    'compress':          'compress',
+    'named_pipe':        'pipe',
+    'read_default_file': 'defaults-file',
+    'charset':           'default-character-set',
+    'local_infile':      'local-infile'
+}
+# Map of mysql executable option names to MySQLdb driver connect() keywords.
+# Note: 'read_default_group' is not supported, since it can cause
+# the MySQLdb driver and the mysql executable to connect differently.
+optionToConnectArgMap = \
+    dict((v,k) for (k,v) in connectArgToOptionMap.iteritems())
+
+def readCredentialFile(fName, logger):
+    """
+    Reads all supported key/value pairs from fName and return a dictionary
+    containing these key/value pairs translated to names accepted by connect()
+    as needed). Hint, to get a subset, do something like:
+    dict = readCredentialFile(fN)
+    (hst, prt, usr, pwd) = [dict[k] for k in ('host', 'port', 'user', 'passwd')].
+    This function only reads from the [mysql] section, e.g., it is not full
+    equivalent to how mysql command like utility which obtains the value from the
+    last occurrence of k in section [mysql] or [client] in the file.
+    """
+    ret = {}
+    if fName.startswith('~'):
+        fName = os.path.expanduser(fName)
+    if not os.path.isfile(fName):
+        raise Exception("Required file '%s' not found" % fName)
+    cnf = ConfigParser.ConfigParser()
```

Comment:
It's not 100% compatible, because mysql config file can include [directives to include other or directories](https://github.com/mysql/mysql-server/blob/7e58ce8e69de5a4e941d913999cbbf905fe8f444/mysys_ssl/my_default.cc#L867). However, a quick glance at the code looks like any .ini file is 100% compatible with a mysql config file and should include enough options to configure any client (which is what we are using the DB module for anyway).

I think the proper way to deal with this is to have people write .ini compliant config files and let an exception pop up if there is an error.